tests/e2e/ppnd-e2e/ppnd_test.go

```diff
@@ -0,0 +1,367 @@
+				},
+			}
+
+			UpdatePipelineRollout(slowPipelineRolloutName, *slowPipelineSpec, numaflowv1.PipelinePhasePausing, func(retrievedPipelineSpec numaflowv1.PipelineSpec) bool {
+				return true
+			}, true, false)
+
+			completeSlowPipelineRolloutTest()
+
+		}
+	})
+
+	It("Should allow data loss in the Pipeline if requested - ISBService update", func() {
+		if UpgradeStrategy == config.PPNDStrategyID {
+
+			createSlowPipelineRollout()
+
+			Document("Updating ISBService to cause a PPND change")
+			updatedISBServiceSpec := isbServiceSpec
+			updatedISBServiceSpec.JetStream.Version = updatedJetstreamVersion
+			rawSpec, err := json.Marshal(updatedISBServiceSpec)
+			Expect(err).ShouldNot(HaveOccurred())
+
+			// we use the direct k8s function since we need to control the PipelineRollout to complete the update
+			UpdateISBServiceRolloutInK8S(isbServiceRolloutName, func(rollout apiv1.ISBServiceRollout) (apiv1.ISBServiceRollout, error) {
+				rollout.Spec.InterStepBufferService.Spec.Raw = rawSpec
+				return rollout, nil
+			})
+
+			completeSlowPipelineRolloutTest()
+
+			// confirm update
+			VerifyISBServiceSpec(Namespace, isbServiceRolloutName, func(retrievedISBServiceSpec numaflowv1.InterStepBufferServiceSpec) bool {
+				return retrievedISBServiceSpec.JetStream.Version == updatedJetstreamVersion
+			})
+
+		}
+	})
+
+	It("Should allow data loss in the Pipeline if requested - Numaflow Controller update", func() {
+		if UpgradeStrategy == config.PPNDStrategyID {
+
+			createSlowPipelineRollout()
+
+			Document("Updating Numaflow controller to cause a PPND change")
+			updatedNumaflowControllerROSpec := apiv1.NumaflowControllerRolloutSpec{
+				Controller: apiv1.Controller{Version: updatedNumaflowControllerVersion},
+			}
+			UpdateNumaflowControllerRolloutInK8S(func(rollout apiv1.NumaflowControllerRollout) (apiv1.NumaflowControllerRollout, error) {
+				rollout.Spec = updatedNumaflowControllerROSpec
+				return rollout, nil
+			})
+
+			completeSlowPipelineRolloutTest()
+
+			// confirm update
+			VerifyNumaflowControllerDeployment(Namespace, func(d appsv1.Deployment) bool {
+				colon := strings.Index(d.Spec.Template.Spec.Containers[0].Image, ":")
+				return colon != -1 && d.Spec.Template.Spec.Containers[0].Image[colon+1:] == "v"+updatedNumaflowControllerVersion
+			})
+
+		}
+	})
+
+	It("Should update a Pipeline even if the Pipeline is failed", func() {
+
+		// add bad edge to automatically fail Pipeline
+		failedPipelineSpec := initialPipelineSpec
+		failedPipelineSpec.Edges = append(failedPipelineSpec.Edges, numaflowv1.Edge{From: "not", To: "valid"})
+
+		CreateFailedPipelineRollout(failedPipelineRolloutName, Namespace, failedPipelineSpec)
+		VerifyPipelineFailed(Namespace, failedPipelineRolloutName)
+
+		time.Sleep(5 * time.Second)
+
+		// update spec to have topology change
+		failedPipelineSpec = updatedPipelineSpec
+		failedPipelineSpec.Edges = append(failedPipelineSpec.Edges, numaflowv1.Edge{From: "not", To: "valid"})
```

Comment:
so, the part that matters for this test is actually the fact that it was failed beforehand - not that the spec afterward was failed. Before the before spec was failed, we did not try to pause and were able to perform the update.

We should be able to simply update to `initialPipelineSpec` even, or otherwise `updatedPipelineSpec`.

Okay to do in a subsequent PR if you prefer.

Reply:
Looking at this again.. I see the issue and will just update to updatedPipelineSpec without the bad edge. I assumed we wanted to keep the Pipeline in a failed state.